eventuals/concurrent-ordered.h

```diff
@@ -0,0 +1,254 @@
+    }
+  };
+};
+
+/////////////////////////////////////////////////////////////////////
+
+template <typename Value>
+inline auto ReorderAdaptor() {
+  return _ReorderAdaptor::Composable<Value>{};
+}
+
+/////////////////////////////////////////////////////////////////////
+
+// Needed to indicate the end of streaming by passing tuple with '-index'.
+// Acting like both of 'Stream' and 'Loop'. In case of propagating each value
+// as tuple in 'Body' and 'Ended' need to indicate if stream is ended, so it
+// saves upstream and on `Next` tries to get a next value from upstream,
+// if not ended, otherwise it calls 'Ended'. So 'Next' is both ending and
+// getting next values function.
+struct _ConcurrentOrderedAdaptor {
+  template <typename K_, typename Value_>
+  struct Continuation : public TypeErasedStream {
+    Continuation(K_ k)
+      : k_(std::move(k)) {}
+
+    void Begin(TypeErasedStream& stream) {
+      upstream_ = &stream;
+      ended_ = false;
+      k_.Begin(*this);
+    }
+
+    template <typename Value>
+    void Body(std::tuple<int, Value>&& tuple) {
+      // NOTE: Either this is the first value we've received on this stream
+      // or the index, should be the same as the value we received before.
+      int i = std::get<0>(tuple);
+
+      CHECK(!index_ || index_.value() == i);
+
+      index_ = i;
+      k_.Body(std::make_tuple(
+          i,
+          std::make_optional(std::move(std::get<1>(tuple)))));
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Register(Interrupt& interrupt) {
+      k_.Register(interrupt);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    void Ended() {
+      ended_ = true;
+      CHECK(index_);
+      k_.Body(std::make_tuple(index_.value() * -1, std::nullopt));
+    }
+
+    void Next() override {
+      if (ended_) {
+        k_.Ended();
+      } else {
+        upstream_->Next();
+      }
+    }
+
+    void Done() override {
+      upstream_->Done();
+    }
+
+    K_ k_;
+
+    bool ended_ = false;
+
+    std::optional<int> index_;
+
+    TypeErasedStream* upstream_;
+  };
+
+  template <typename Value_>
+  struct Composable {
+    template <typename>
+    using ValueFrom = std::tuple<int, std::optional<Value_>>;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Value_>(std::move(k));
+    }
+  };
+};
+
+/////////////////////////////////////////////////////////////////////
+
+template <typename Value>
+inline auto ConcurrentOrderedAdaptor() {
+  return _ConcurrentOrderedAdaptor::Composable<Value>{};
+}
+
+/////////////////////////////////////////////////////////////////////
+
+} // namespace detail
+
+/////////////////////////////////////////////////////////////////////
+
+// Specify return type 'T' for 'ConcurrentOrdered()' to pass it into
+// 'ConcurrentOrderedAdaptor()' and 'ReorderAdaptor()'.
+template <typename T, typename F>
```

Comment:
We shouldn't have to have the return type `T`. You should be able to get the return type from `f`.